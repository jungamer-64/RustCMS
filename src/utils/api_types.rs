use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use axum::Json;
use serde_json::{json, Value};
use utoipa::ToSchema;

/// 標準的なAPIレスポンス構造
///
/// ジェネリック型 `T` を含むラッパー。OpenAPI 上では良く使う組合せを aliases で公開する。
/// 例: `ApiResponsePostResponse` は `ApiResponse<PostResponse>` のエイリアス。
#[derive(Debug, Serialize, ToSchema)]
#[schema(example = json!({
    "success": true,
    "data": {"example": "value"},
    "message": null,
    "error": null,
    "validation_errors": null
}))]
pub struct ApiResponse<T> {
    pub success: bool,
    pub data: Option<T>,
    pub message: Option<String>,
    pub error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub validation_errors: Option<Vec<ValidationError>>,    
}

impl<T> ApiResponse<T> {
    pub fn success(data: T) -> Self {
        Self {
            success: true,
            data: Some(data),
            message: None,
            error: None,
            validation_errors: None,
        }
    }

    pub fn success_with_message(data: T, message: String) -> Self {
        Self {
            success: true,
            data: Some(data),
            message: Some(message),
            error: None,
            validation_errors: None,
        }
    }

}

impl ApiResponse<()> {
    pub fn error(error: String) -> Self {
        Self {
            success: false,
            data: None,
            message: None,
            error: Some(error),
            validation_errors: None,
        }
    }

    pub fn error_with_validation(error: String, validation_errors: Vec<ValidationError>) -> Self {
        Self {
            success: false,
            data: None,
            message: None,
            error: Some(error),
            validation_errors: Some(validation_errors),
        }
    }
}

/// 便利用成功ヘルパ: `Json(ApiResponse::success(data))` の重複削減。
#[deprecated(note = "Use ApiOk(value) or Json(ApiResponse::success(value)) へ移行。Ok ヘルパは今後削除予定。")]
pub fn ok<T: Serialize>(data: T) -> Json<ApiResponse<T>> {
    Json(ApiResponse::success(data))
}

/// メッセージのみの成功レスポンス (data は空 JSON オブジェクト扱い)。
#[deprecated(note = "Use ApiOk(json!({ \"message\": ... })) もしくは Json(ApiResponse::success(...)) へ移行。")]
pub fn ok_message(msg: impl Into<String>) -> Json<ApiResponse<Value>> {
    Json(ApiResponse::success(json!({ "message": msg.into() })))
}

/// エラーヘルパ: `Json(ApiResponse::error(msg))` の重複削減。
#[deprecated(note = "Use Json(ApiResponse::error(...)) へ移行。err ヘルパは今後削除予定。")]
pub fn err(msg: impl Into<String>) -> Json<ApiResponse<()>> {
    let resp: ApiResponse<()> = ApiResponse::error(msg.into());
    Json(resp)
}

/// ページネーション情報
#[derive(Debug, Serialize, Deserialize, ToSchema)]
#[schema(example = json!({"page":1,"per_page":20,"total":42,"total_pages":3}))]
pub struct Pagination {
    pub page: u32,
    pub per_page: u32,
    pub total: u64,
    pub total_pages: u32,
}

/// ページネーション付きレスポンス
#[derive(Debug, Serialize, ToSchema)]
pub struct PaginatedResponse<T> {
    pub data: Vec<T>,
    pub pagination: Pagination,
}

/// ページネーションクエリパラメータ
#[derive(Debug, Deserialize, ToSchema)]
pub struct PaginationQuery {
    #[serde(default = "default_page")]
    pub page: u32,
    #[serde(default = "default_per_page")]
    pub per_page: u32,
}

fn default_page() -> u32 {
    1
}
fn default_per_page() -> u32 {
    20
}

impl Default for PaginationQuery {
    fn default() -> Self {
        Self {
            page: 1,
            per_page: 20,
        }
    }
}

impl PaginationQuery {
    pub fn validate(&mut self) {
        if self.page == 0 {
            self.page = 1;
        }
        if self.per_page == 0 {
            self.per_page = 20;
        }
        if self.per_page > 100 {
            self.per_page = 100;
        }
    }

    pub fn offset(&self) -> u64 {
        ((self.page - 1) * self.per_page) as u64
    }
}

/// ソートパラメータ
#[derive(Debug, Deserialize, ToSchema)]
pub struct SortQuery {
    #[serde(default)]
    pub sort_by: Option<String>,
    #[serde(default)]
    pub order: SortOrder,
}

#[derive(Debug, Deserialize, Serialize, ToSchema, Default)]
#[serde(rename_all = "lowercase")]
pub enum SortOrder {
    Asc,
    #[default]
    Desc,
}

/// フィルタークエリパラメータ
#[derive(Debug, Deserialize, ToSchema)]
pub struct FilterQuery {
    pub search: Option<String>,
    pub status: Option<String>,
    pub category: Option<String>,
    pub tag: Option<String>,
}

/// リソース作成レスポンス
#[derive(Debug, Serialize, ToSchema)]
pub struct CreatedResponse {
    pub id: String,
    pub message: String,
    pub created_at: DateTime<Utc>,
}

/// リソース更新レスポンス
#[derive(Debug, Serialize, ToSchema)]
pub struct UpdatedResponse {
    pub id: String,
    pub message: String,
    pub updated_at: DateTime<Utc>,
}

/// リソース削除レスポンス
#[derive(Debug, Serialize, ToSchema)]
pub struct DeletedResponse {
    pub id: String,
    pub message: String,
    pub deleted_at: DateTime<Utc>,
}

/// バリデーションエラー
#[derive(Debug, Serialize, ToSchema)]
#[schema(example = json!({"field":"title","message":"must not be empty"}))]
pub struct ValidationError {
    pub field: String,
    pub message: String,
}

// ValidationErrorResponse は統合のため削除 (ApiResponse に統合)

/// ApiResponse の具体例（ジェネリックなし）
#[derive(Debug, Serialize, ToSchema)]
#[schema(example = json!({
    "success": true,
    "data": {"message": "Operation completed"},
    "message": null,
    "error": null,
    "validation_errors": null
}))]
pub struct ApiResponseExample {
    pub success: bool,
    pub data: Option<Value>,
    pub message: Option<String>,
    pub error: Option<String>,
    pub validation_errors: Option<Vec<ValidationError>>,
}
