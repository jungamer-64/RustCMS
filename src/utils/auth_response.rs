use crate::utils::common_types::UserInfo;
use serde::Serialize;
use utoipa::ToSchema;

#[derive(Debug, Serialize, ToSchema, Clone)]
pub struct AuthTokens {
    pub access_token: String,
    pub refresh_token: String,
    pub biscuit_token: String,
    pub expires_in: i64,
    pub session_id: String,
}

#[derive(Debug, Serialize)]
pub struct AuthSuccess<T> {
    pub success: bool,
    pub tokens: AuthTokens,
    pub user: Option<UserInfo>,
    pub extra: Option<T>,
}

impl<T> AuthSuccess<T> {
    #[must_use]
    pub const fn new(tokens: AuthTokens, user: Option<UserInfo>, extra: Option<T>) -> Self {
        Self {
            success: true,
            tokens,
            user,
            extra,
        }
    }
}

impl From<crate::auth::AuthResponse> for AuthTokens {
    fn from(a: crate::auth::AuthResponse) -> Self {
        a.tokens
    }
}

/// 統一認証レスポンス (login/register 用)
///
/// Biscuit トークンベースの認証情報を `tokens` オブジェクトで提供します。
#[derive(Debug, Serialize, ToSchema)]
pub struct AuthSuccessResponse {
    pub success: bool,
    pub tokens: AuthTokens,
    pub user: UserInfo,
}

impl From<crate::auth::AuthResponse> for AuthSuccessResponse {
    fn from(a: crate::auth::AuthResponse) -> Self {
        Self::from_parts(&a.tokens, a.user)
    }
}

impl AuthSuccessResponse {
    #[must_use]
    pub fn from_parts(tokens: &AuthTokens, user: UserInfo) -> Self {
        Self {
            success: true,
            tokens: tokens.clone(),
            user,
        }
    }
}
