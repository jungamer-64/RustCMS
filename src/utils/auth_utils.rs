//! Authentication utilities
//!
//! Common authentication helper functions

#![allow(deprecated)] // internal deprecations intentionally retained under feature gate

#[cfg(feature = "legacy-admin-token")]
use std::env;

/// Check if the provided admin token is valid
/// This is a simple token-based authentication for admin endpoints
/// 
/// # Deprecated
/// This function is deprecated in favor of Biscuit-based authentication with role permissions.
/// Use the unified authentication system with "admin" permission checking instead.
#[cfg(feature = "legacy-admin-token")]
#[deprecated(note = "Use Biscuit authentication with admin permission checking instead (will be removed in 3.0.0)")]
pub fn check_admin_token(req_token: &str) -> bool {
    env::var("ADMIN_TOKEN")
        .map(|t| t == req_token)
        .unwrap_or(false)
}

/// Get the admin token from environment
/// 
/// # Deprecated
/// This function is deprecated in favor of Biscuit-based authentication with role permissions.
/// Use the unified authentication system with "admin" permission checking instead.
#[cfg(feature = "legacy-admin-token")]
#[deprecated(note = "Use Biscuit authentication with admin permission checking instead (will be removed in 3.0.0)")]
pub fn get_admin_token() -> Option<String> {
    env::var("ADMIN_TOKEN").ok()
}

#[cfg(all(test, feature = "legacy-admin-token"))]
mod tests {
    use super::*;
    use std::env;
    use std::sync::Mutex;

    // Mutex to ensure tests run sequentially to avoid env var conflicts
    static TEST_MUTEX: Mutex<()> = Mutex::new(());

    #[cfg(feature = "legacy-admin-token")]
    #[test]
    fn test_check_admin_token() {
        let _lock = TEST_MUTEX.lock().unwrap();

        // Save original value
        let original = env::var("ADMIN_TOKEN").ok();

        // Test with no environment variable
        env::remove_var("ADMIN_TOKEN");
        assert!(!check_admin_token("any_token"));

        // Test with matching token
        env::set_var("ADMIN_TOKEN", "test_token");
        assert!(check_admin_token("test_token"));

        // Test with non-matching token
        assert!(!check_admin_token("wrong_token"));

        // Restore original value
        match original {
            Some(val) => env::set_var("ADMIN_TOKEN", val),
            None => env::remove_var("ADMIN_TOKEN"),
        }
    }

    #[cfg(feature = "legacy-admin-token")]
    #[test]
    fn test_check_admin_token_empty() {
        let _lock = TEST_MUTEX.lock().unwrap();

        // Save original value
        let original = env::var("ADMIN_TOKEN").ok();

        // Test with empty token in environment
        env::set_var("ADMIN_TOKEN", "");
        assert!(check_admin_token(""));
        assert!(!check_admin_token("not_empty"));

        // Restore original value
        match original {
            Some(val) => env::set_var("ADMIN_TOKEN", val),
            None => env::remove_var("ADMIN_TOKEN"),
        }
    }

    #[cfg(feature = "legacy-admin-token")]
    #[test]
    fn test_get_admin_token() {
        let _lock = TEST_MUTEX.lock().unwrap();

        // Save original value
        let original = env::var("ADMIN_TOKEN").ok();

        // Test with no environment variable
        env::remove_var("ADMIN_TOKEN");
        assert!(get_admin_token().is_none());

        // Test with set token
        env::set_var("ADMIN_TOKEN", "test_admin_token");
        assert_eq!(get_admin_token(), Some("test_admin_token".to_string()));

        // Test with empty token
        env::set_var("ADMIN_TOKEN", "");
        assert_eq!(get_admin_token(), Some("".to_string()));

        // Restore original value
        match original {
            Some(val) => env::set_var("ADMIN_TOKEN", val),
            None => env::remove_var("ADMIN_TOKEN"),
        }
    }

    #[cfg(feature = "legacy-admin-token")]
    #[test]
    fn test_check_admin_token_special_characters() {
        let _lock = TEST_MUTEX.lock().unwrap();

        // Save original value
        let original = env::var("ADMIN_TOKEN").ok();

        let special_token = "token!@#$%^&*(){}[]|\\:;\"'<>,.?/~`";
        env::set_var("ADMIN_TOKEN", special_token);

        assert!(check_admin_token(special_token));
        assert!(!check_admin_token("different_token"));

        // Restore original value
        match original {
            Some(val) => env::set_var("ADMIN_TOKEN", val),
            None => env::remove_var("ADMIN_TOKEN"),
        }
    }

    #[cfg(feature = "legacy-admin-token")]
    #[test]
    fn test_check_admin_token_unicode() {
        let _lock = TEST_MUTEX.lock().unwrap();

        // Save original value
        let original = env::var("ADMIN_TOKEN").ok();

        let unicode_token = "ãƒˆãƒ¼ã‚¯ãƒ³123ðŸ”‘";
        env::set_var("ADMIN_TOKEN", unicode_token);

        assert!(check_admin_token(unicode_token));
        assert!(!check_admin_token("regular_token"));

        // Restore original value
        match original {
            Some(val) => env::set_var("ADMIN_TOKEN", val),
            None => env::remove_var("ADMIN_TOKEN"),
        }
    }
}
