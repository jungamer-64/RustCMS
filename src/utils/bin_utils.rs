use axum::{Router, routing::get};
use std::net::SocketAddr;

/// Print environment summary and recommended settings for local runs
pub fn print_env_summary() {
    use std::env;

    println!("Environment variables for CMS:");
    println!(
        "DATABASE_URL: {}",
        env::var("DATABASE_URL").unwrap_or_else(|_| "Not set".to_string())
    );
    println!(
        "DATABASE_NAME: {}",
        env::var("DATABASE_NAME").unwrap_or_else(|_| "Not set".to_string())
    );
    println!(
        "SERVER_HOST: {}",
        env::var("SERVER_HOST").unwrap_or_else(|_| "Not set".to_string())
    );
    println!(
        "SERVER_PORT: {}",
        env::var("SERVER_PORT").unwrap_or_else(|_| "Not set".to_string())
    );

    // 推奨設定を表示
    println!("\n推奨設定を表示");
    println!("DATABASE_URL=postgres://user:pass@localhost:5432/rust_cms");
    println!("DATABASE_NAME=rust_cms");
    println!("SERVER_HOST=127.0.0.1");
    println!("SERVER_PORT=3001");
    println!("BISCUIT_ROOT_KEY=base64_or_path_to_keydir");

    println!("\n✅ To run the CMS, use:");
    println!("cargo run --bin cms-server");
}

/// Render a health table from primitive components for CLI output.
/// Public so binaries and tests can reuse the same rendering logic.
pub fn render_health_table_components(
    overall_status: &str,
    db: (&str, f64, Option<&str>),
    cache: (&str, f64, Option<&str>),
    search: (&str, f64, Option<&str>),
    auth: (&str, f64, Option<&str>),
) -> comfy_table::Table {
    use comfy_table::{Cell, Table};

    let mut table = Table::new();
    table.set_header(vec!["Component", "Status", "Response (ms)", "Error"]);

    table.add_row(vec![
        Cell::new("Overall"),
        Cell::new(overall_status.to_string()),
        Cell::new("-"),
        Cell::new("-"),
    ]);

    table.add_row(vec![
        Cell::new("Database"),
        Cell::new(db.0.to_string()),
        Cell::new(format!("{:.2}", db.1)),
        Cell::new(db.2.unwrap_or_default()),
    ]);

    table.add_row(vec![
        Cell::new("Cache"),
        Cell::new(cache.0.to_string()),
        Cell::new(format!("{:.2}", cache.1)),
        Cell::new(cache.2.unwrap_or_default()),
    ]);

    table.add_row(vec![
        Cell::new("Search"),
        Cell::new(search.0.to_string()),
        Cell::new(format!("{:.2}", search.1)),
        Cell::new(search.2.unwrap_or_default()),
    ]);

    table.add_row(vec![
        Cell::new("Auth"),
        Cell::new(auth.0.to_string()),
        Cell::new(format!("{:.2}", auth.1)),
        Cell::new(auth.2.unwrap_or_default()),
    ]);

    table
}

/// Build a minimal docs-only Router that doesn't require `AppState`
fn docs_router() -> Router {
    use crate::handlers;
    Router::new()
        .route("/api/docs", get(handlers::docs_ui))
        .route("/api/docs/openapi.json", get(handlers::openapi_json))
}

/// Run a lightweight docs server on the given address.
///
/// # Errors
///
/// バインド時にソケットの確保に失敗した場合や、サーバー実行中に内部エラーが発生した場合にエラーを返します。
pub async fn run_docs_server(addr: SocketAddr) -> crate::Result<()> {
    let app = docs_router();

    println!("Docs server running on http://{addr} (endpoints: /api/docs, /api/docs/openapi.json)");

    let listener = tokio::net::TcpListener::bind(addr)
        .await
        .map_err(crate::AppError::IO)?;
    axum::serve(listener, app)
        .await
        .map_err(|e| crate::AppError::Internal(format!("axum serve error: {e}")))?;
    Ok(())
}
