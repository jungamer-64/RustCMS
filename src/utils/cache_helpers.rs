use std::time::Duration;

/// Cache-or-compute helper that hides cache feature flag branching.
///
/// If the `cache` feature is enabled it will first attempt to read from cache,
/// otherwise it computes the value and stores it with the provided TTL.
/// When the `cache` feature is disabled, it simply computes and returns the value.
///
/// # Errors
///
/// Returns an error if the `compute` future returns an error. Cache get/set
/// failures are ignored on set (best-effort) but a successful cache hit is
/// required to deserialize into `T`; deserialization failures will propagate
/// as an error from the underlying cache client.
pub async fn cache_or_compute<T, F, Fut>(
    state: crate::AppState,
    key: &str,
    ttl_secs: u64,
    compute: F,
) -> crate::Result<T>
where
    T: serde::de::DeserializeOwned + serde::Serialize + Send + Sync + 'static,
    F: FnOnce() -> Fut,
    Fut: std::future::Future<Output = crate::Result<T>>,
{
    #[cfg(feature = "cache")]
    {
        if let Ok(Some(v)) = state.cache.get::<T>(key).await {
            return Ok(v);
        }
        let v = compute().await?;
        // best-effort set
        let _ = state
            .cache
            .set(key.to_string(), &v, Some(Duration::from_secs(ttl_secs)))
            .await;
        Ok(v)
    }
    #[cfg(not(feature = "cache"))]
    {
        compute().await
    }
}
