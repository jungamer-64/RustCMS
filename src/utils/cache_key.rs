//! Cache key builder utility to avoid ad-hoc format! duplication.
//! Produces colon separated stable keys: namespace + labeled segments.
//! Example: CacheKeyBuilder::new("posts")
//!   .kv("page", 1).kv("limit", 20).kv_opt("status", None)
//!   .build() => "posts:page:1:limit:20:status:all".

pub struct CacheKeyBuilder {
    base: String,
    segs: Vec<String>,
    // keep track of label parts to avoid accidental duplicates (debug only enforcement)
    used_labels: std::collections::HashSet<String>,
}

// Common cache key base & prefixes (keep centralized for invalidation & consistency)
pub const CACHE_PREFIX_POST_ID: &str = "post:id:"; // + {uuid}
pub const CACHE_PREFIX_POSTS: &str = "posts:"; // list queries start with this
pub const CACHE_PREFIX_USER_ID: &str = "user:id:"; // + {uuid}
pub const CACHE_PREFIX_USERS: &str = "users:"; // list queries start with this
pub const CACHE_PREFIX_USER_POSTS: &str = "user_posts:user:"; // + {uuid}:...

/// Convenience helper to build a simple entity id based cache key.
/// Example: entity_id_key("post", uuid) => "post:id:{uuid}"
pub fn entity_id_key(prefix: &str, id: impl std::fmt::Display) -> String {
    format!("{}:id:{}", prefix, id)
}

/// Enum describing list style cache keys we standardize on. This avoids a
/// proliferation of small one-off wrapper functions like build_posts_cache_key
/// or build_users_cache_key. Handlers can format through `to_cache_key`.
pub enum ListCacheKey<'a> {
    Posts {
        page: u32,
        limit: u32,
        status: &'a Option<String>,
        author: &'a Option<uuid::Uuid>,
        tag: &'a Option<String>,
        sort: &'a Option<String>,
    },
    Users {
        page: u32,
        limit: u32,
        role: &'a Option<String>,
        active: Option<bool>,
        sort: &'a Option<String>,
    },
}

impl<'a> ListCacheKey<'a> {
    pub fn to_cache_key(&self) -> String {
        match self {
            ListCacheKey::Posts {
                page,
                limit,
                status,
                author,
                tag,
                sort,
            } => build_list_cache_key(
                "posts",
                *page,
                *limit,
                &[
                    ("status", (*status).clone()),
                    ("author", author.map(|u| u.to_string())),
                    ("tag", (*tag).clone()),
                    ("sort", (*sort).clone()),
                ],
            ),
            ListCacheKey::Users {
                page,
                limit,
                role,
                active,
                sort,
            } => build_list_cache_key(
                "users",
                *page,
                *limit,
                &[
                    ("role", (*role).clone()),
                    ("active", active.map(|b| b.to_string())),
                    ("sort", (*sort).clone()),
                ],
            ),
        }
    }
}

impl CacheKeyBuilder {
    pub fn new(base: impl Into<String>) -> Self {
        Self {
            base: base.into(),
            segs: Vec::new(),
            used_labels: std::collections::HashSet::new(),
        }
    }
    fn push_kv(&mut self, key: &str, val: String) {
        // Enforce uniqueness of labels to prevent accidental overwrites like .kv("page",1).kv("page",2)
        debug_assert!(
            !self.used_labels.contains(key),
            "duplicate cache key segment label detected: {key}"
        );
        self.used_labels.insert(key.to_string());
        self.segs.push(format!("{}:{}", key, val));
    }
    pub fn kv(mut self, key: &str, value: impl std::fmt::Display) -> Self {
        self.push_kv(key, value.to_string());
        self
    }
    pub fn kv_opt<T: std::fmt::Display>(mut self, key: &str, opt: Option<T>) -> Self {
        match opt {
            Some(v) => self.push_kv(key, v.to_string()),
            None => self.push_kv(key, "all".to_string()),
        }
        self
    }
    pub fn build(self) -> String {
        if self.segs.is_empty() {
            self.base
        } else {
            format!("{}:{}", self.base, self.segs.join(":"))
        }
    }
}

/// Helper to build a common list-style cache key used by many handlers.
///
/// Accepts a base (e.g. "posts" or "users"), page/limit and a slice of
/// (label, Option<String>) pairs for additional labeled segments. When the
/// Option is None the key will encode the segment as "label:all" using
/// `kv_opt` semantics.
pub fn build_list_cache_key(
    base: &str,
    page: u32,
    limit: u32,
    pairs: &[(&str, Option<String>)],
) -> String {
    let mut b = CacheKeyBuilder::new(base)
        .kv("page", page)
        .kv("limit", limit);
    for (k, v) in pairs {
        match v {
            Some(val) => {
                b = b.kv(k, val);
            }
            None => {
                b = b.kv_opt::<String>(k, None);
            }
        }
    }
    b.build()
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn builds_expected() {
        let k = CacheKeyBuilder::new("posts")
            .kv("page", 1)
            .kv("limit", 20)
            .kv_opt("status", Option::<String>::None)
            .build();
        assert_eq!(k, "posts:page:1:limit:20:status:all");
    }

    #[test]
    fn ordering_is_stable() {
        let a = CacheKeyBuilder::new("x").kv("b", 2).kv("a", 1).build();
        assert_eq!(a, "x:b:2:a:1"); // insertion order preserved
    }

    #[test]
    fn unicode_values_supported() {
        let k = CacheKeyBuilder::new("tag").kv("名前", "値").build();
        assert_eq!(k, "tag:名前:値");
    }

    #[test]
    #[should_panic(expected = "duplicate cache key segment label detected")]
    fn duplicate_label_panics_in_debug() {
        let _ = CacheKeyBuilder::new("dup")
            .kv("page", 1)
            .kv("page", 2)
            .build();
    }
}
