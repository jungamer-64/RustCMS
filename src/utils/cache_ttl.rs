//! Centralized cache TTL constants
pub const CACHE_TTL_SHORT: u64 = 120;   // suggestions, transient
pub const CACHE_TTL_DEFAULT: u64 = 300; // standard lists and lookups
pub const CACHE_TTL_LONG: u64 = 600;    // user/profile like
