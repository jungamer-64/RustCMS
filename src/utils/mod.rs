//! Utility modules (Phase 7 minimal set - legacy removed)

pub mod api_types;
pub mod cache_key;
pub mod cache_ttl;
pub mod date;
pub mod deprecation;
// Phase 9: dto module removed (legacy code deleted in Phase 7)
// pub mod dto;
pub mod dup;
// pub mod error; // Removed in Phase 7
pub mod file;
pub mod hash;
pub mod password;
pub mod response_ext;
// Phase 9: search_index module removed (legacy code deleted in Phase 7)
// pub mod search_index;
pub mod security_validation;
pub mod sort;
pub mod text;
pub mod url_encoding;
pub mod validation;
pub mod vec_helpers;
