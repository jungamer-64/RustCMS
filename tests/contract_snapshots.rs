use cms_backend::utils::api_types::{ApiResponse, ValidationError};
use serde_json::json;
use insta::assert_json_snapshot;

// 初期スナップショット: 統一レスポンス構造 (成功)
#[test]
fn snapshot_api_success_basic() {
    let resp = ApiResponse::success(json!({"foo":"bar","num":123}));
    assert_json_snapshot!("api_success_basic", resp);
    assert!(resp.success);
}

// メッセージ付き成功 (回帰検知用)
#[test]
fn snapshot_api_success_with_message() {
    let resp = ApiResponse::success_with_message(json!({"ok":true}), "done".to_string());
    assert_json_snapshot!("api_success_with_message", resp);
    assert!(resp.success);
}

// エラーパターン (バリデーションなし)
#[test]
fn snapshot_api_error_basic() {
    let resp = ApiResponse::error("something went wrong".to_string());
    assert_json_snapshot!("api_error_basic", resp);
    assert!(!resp.success);
}

// バリデーションエラー付き
#[test]
fn snapshot_api_error_with_validation() {
    let resp = ApiResponse::error_with_validation(
        "validation failed".to_string(),
        vec![
            ValidationError { field: "email".into(), message: "invalid format".into() },
            ValidationError { field: "password".into(), message: "too short".into() },
        ]
    );
    assert_json_snapshot!("api_error_with_validation", resp);
    assert!(!resp.success);
}
