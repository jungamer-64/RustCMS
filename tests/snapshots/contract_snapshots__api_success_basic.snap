---
source: tests/contract_snapshots.rs
expression: resp
---
{
  "success": true,
  "data": {
    "foo": "bar",
    "num": 123
  },
  "message": null,
  "error": null
}
